File: backend/routes/stickerRoutes.js
const express = require('express');
const router = express.Router();
const { protect } = require('../middleware/auth');
const Sticker = require('../models/Sticker');
const multer = require('multer');
const { CloudinaryStorage } = require('multer-storage-cloudinary');
const cloudinary = require('cloudinary').v2;

// ─────────────────────────────────────────────────────────────
// Configure Cloudinary
// FIX: was CLOUDINARY_NAME — must match backend/.env key exactly
// ─────────────────────────────────────────────────────────────
cloudinary.config({
  cloud_name: process.env.CLOUDINARY_CLOUD_NAME,   // ← fixed
  api_key:    process.env.CLOUDINARY_API_KEY,
  api_secret: process.env.CLOUDINARY_API_SECRET,
});

const storage = new CloudinaryStorage({
  cloudinary,
  params: {
    folder: 'prabha_stickers',
    allowed_formats: ['jpg', 'png', 'jpeg', 'webp'],
  },
});

const upload = multer({
  storage,
  limits: { fileSize: 5 * 1024 * 1024 }, // 5 MB max
});

// ─────────────────────────────────────────────────────────────
// Allowed category values (whitelist)
// ─────────────────────────────────────────────────────────────
const ALLOWED_CATEGORIES = [
  'Car and Bike Stickers',
  'Business Signage',
  'Specialty Items',
  'General',
];

// @desc    Get all stickers
// @route   GET /api/stickers
// @access  Public
router.get('/', async (req, res) => {
  try {
    const stickers = await Sticker.find().sort({ createdAt: -1 });
    res.json(stickers);
  } catch (error) {
    console.error('GET /stickers error:', error.message);
    res.status(500).json({ message: 'Server Error' });
  }
});

// @desc    Create a sticker
// @route   POST /api/stickers
// @access  Private
router.post('/', protect, upload.single('image'), async (req, res) => {
  try {
    if (!req.file) {
      return res.status(400).json({ message: 'No image provided' });
    }

    // ── Input validation ──────────────────────────────────────
    const title    = (req.body.title    || '').trim().slice(0, 100) || 'Untitled';
    const category = (req.body.category || '').trim();

    if (category && !ALLOWED_CATEGORIES.includes(category)) {
      return res.status(400).json({
        message: `Invalid category. Allowed values: ${ALLOWED_CATEGORIES.join(', ')}`,
      });
    }

    const sticker = new Sticker({
      title,
      imageUrl:  req.file.path,
      publicId:  req.file.filename,
      category:  ALLOWED_CATEGORIES.includes(category) ? category : 'General',
    });

    const created = await sticker.save();
    res.status(201).json(created);
  } catch (error) {
    console.error('POST /stickers error:', error.message);
    res.status(500).json({ message: 'Server Error' });
  }
});

// @desc    Delete a sticker
// @route   DELETE /api/stickers/:id
// @access  Private
router.delete('/:id', protect, async (req, res) => {
  try {
    const sticker = await Sticker.findById(req.params.id);

    if (!sticker) {
      return res.status(404).json({ message: 'Sticker not found' });
    }

    if (sticker.publicId) {
      await cloudinary.uploader.destroy(sticker.publicId);
    }
    await sticker.deleteOne();
    res.json({ message: 'Sticker removed' });
  } catch (error) {
    console.error('DELETE /stickers/:id error:', error.message);
    res.status(500).json({ message: 'Server Error' });
  }
});

module.exports = router;

File: backend/routes/sticker_routes.py
import os
import datetime
import cloudinary
import cloudinary.uploader
from flask import Blueprint, request, jsonify
from bson import ObjectId
from db import db, serialize_doc
from middleware.auth import protect

sticker_blueprint = Blueprint('sticker_blueprint', __name__)

ALLOWED_IMAGE_EXTENSIONS = {'jpg', 'jpeg', 'png', 'webp'}

# Configure Cloudinary
cloudinary.config(
    cloud_name=os.getenv('CLOUDINARY_CLOUD_NAME'),
    api_key=os.getenv('CLOUDINARY_API_KEY'),
    api_secret=os.getenv('CLOUDINARY_API_SECRET')
)

def cloudinary_is_configured():
    return all([
        os.getenv('CLOUDINARY_CLOUD_NAME'),
        os.getenv('CLOUDINARY_API_KEY'),
        os.getenv('CLOUDINARY_API_SECRET')
    ])

def allowed_image(filename):
    return (
        '.' in filename
        and filename.rsplit('.', 1)[1].lower() in ALLOWED_IMAGE_EXTENSIONS
    )

@sticker_blueprint.route('/', methods=['GET'])
def get_stickers():
    """
    @desc    Get all stickers
    @route   GET /api/stickers
    @access  Public
    """
    if db is None:
        return jsonify({'message': 'Database not connected'}), 500
        
    try:
        # Sort by createdAt descending (-1)
        stickers_cursor = db.stickers.find().sort('createdAt', -1)
        stickers_list = list(stickers_cursor)
        return jsonify(serialize_doc(stickers_list))
    except Exception as e:
        print(f"Error getting stickers: {e}")
        return jsonify({'message': 'Server Error'}), 500

@sticker_blueprint.route('/', methods=['POST'])
@protect
def create_sticker():
    """
    @desc    Create a sticker
    @route   POST /api/stickers
    @access  Private (Guarded by @protect)
    """
    if db is None:
        return jsonify({'message': 'Database not connected'}), 500

    if not cloudinary_is_configured():
        return jsonify({
            'message': 'Cloudinary is not configured. Add CLOUDINARY_CLOUD_NAME, CLOUDINARY_API_KEY, and CLOUDINARY_API_SECRET.'
        }), 500
        
    try:
        title = request.form.get('title', 'Untitled')
        category = request.form.get('category', 'General')
        
        # In Flask, multipart/form-data files are in request.files
        if 'image' not in request.files:
            return jsonify({'message': 'No image provided'}), 400
            
        image_file = request.files['image']
        
        if image_file.filename == '':
            return jsonify({'message': 'No image provided'}), 400

        if not allowed_image(image_file.filename):
            return jsonify({'message': 'Only JPG, PNG, and WEBP images are allowed'}), 400
            
        # Upload the file to Cloudinary directly from Flask's file stream
        upload_result = cloudinary.uploader.upload(
            image_file.stream,
            folder='prabha_stickers',
            allowed_formats=['jpg', 'png', 'jpeg', 'webp']
        )

        if not upload_result.get('secure_url') or not upload_result.get('public_id'):
            return jsonify({'message': 'Cloudinary upload failed'}), 502
        
        now = datetime.datetime.utcnow()
        new_sticker = {
            'title': title,
            'imageUrl': upload_result.get('secure_url'),
            'publicId': upload_result.get('public_id'),
            'category': category,
            'createdAt': now,
            'updatedAt': now
        }
        
        result = db.stickers.insert_one(new_sticker)
        new_sticker['_id'] = result.inserted_id
        
        return jsonify(serialize_doc(new_sticker)), 201
    except Exception as e:
        print(f"Error creating sticker: {e}")
        return jsonify({'message': f'Upload failed: {str(e)}'}), 500

@sticker_blueprint.route('/<sticker_id>', methods=['DELETE'])
@protect
def delete_sticker(sticker_id):
    """
    @desc    Delete a sticker
    @route   DELETE /api/stickers/:id
    @access  Private (Guarded by @protect)
    """
    if db is None:
        return jsonify({'message': 'Database not connected'}), 500
        
    try:
        # Convert path string ID to BSON ObjectId
        try:
            oid = ObjectId(sticker_id)
        except Exception:
            return jsonify({'message': 'Sticker not found'}), 404
            
        sticker = db.stickers.find_one({'_id': oid})
        
        if sticker:
            # Delete from Cloudinary if publicId exists
            public_id = sticker.get('publicId')
            if public_id:
                try:
                    cloudinary.uploader.destroy(public_id)
                except Exception as ex:
                    print(f"Warning: Cloudinary destroy failed: {ex}")
            
            # Delete from database
            db.stickers.delete_one({'_id': oid})
            return jsonify({'message': 'Sticker removed'})
        else:
            return jsonify({'message': 'Sticker not found'}), 404
    except Exception as e:
        print(f"Error deleting sticker: {e}")
        return jsonify({'message': 'Server Error'}), 500
